File: agenkit-cpp/src/infrastructure/retry_enhanced.cpp
#include "agenkit/infrastructure/retry_enhanced.hpp"
#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace agenkit {
namespace infrastructure {

EnhancedRetryConfig::EnhancedRetryConfig()
    : max_attempts(3)
    , initial_backoff(1000)
    , max_backoff(30000)
    , backoff_multiplier(2.0)
    , jitter_type(JitterType::Full)
    , jitter_min_ratio(0.5)
    , enable_budget(false)
    , max_cost_per_hour(100.0)
    , max_retries_per_hour(1000)
    , enable_backpressure(true)
    , backpressure_threshold(0.5)
    , backpressure_window(100)
{
    // Default error strategies
    error_strategies[ErrorClass::Transient] = ErrorStrategy{
        ErrorClass::Transient,
        5,
        std::chrono::milliseconds(100),
        std::chrono::milliseconds(5000),
        2.0,
        true
    };

    error_strategies[ErrorClass::RateLimit] = ErrorStrategy{
        ErrorClass::RateLimit,
        10,
        std::chrono::milliseconds(60000),
        std::chrono::milliseconds(300000),
        1.5,
        true
    };

    error_strategies[ErrorClass::Timeout] = ErrorStrategy{
        ErrorClass::Timeout,
        3,
        std::chrono::milliseconds(2000),
        std::chrono::milliseconds(30000),
        2.0,
        true
    };

    error_strategies[ErrorClass::ServerError] = ErrorStrategy{
        ErrorClass::ServerError,
        3,
        std::chrono::milliseconds(5000),
        std::chrono::milliseconds(60000),
        2.0,
        true
    };

    error_strategies[ErrorClass::ClientError] = ErrorStrategy{
        ErrorClass::ClientError,
        1,
        std::chrono::milliseconds(0),
        std::chrono::milliseconds(0),
        1.0,
        false
    };
}

EnhancedRetryDecorator::EnhancedRetryDecorator(
    std::shared_ptr<Agent> agent,
    const EnhancedRetryConfig& config
) : agent_(std::move(agent))
  , config_(config)
  , rng_(std::random_device{}())
{
    budget_.max_cost = config_.max_cost_per_hour;
    budget_.current_cost = 0.0;
    budget_.max_retries_per_hour = config_.max_retries_per_hour;
    budget_.retry_count = 0;
    budget_.window_start = std::chrono::steady_clock::now();
}

std::string EnhancedRetryDecorator::name() const {
    return agent_->name();
}

std::vector<std::string> EnhancedRetryDecorator::capabilities() const {
    return agent_->capabilities();
}

ErrorClass EnhancedRetryDecorator::classify_error(const std::exception& error) const {
    if (config_.error_classifier) {
        return config_.error_classifier(error);
    }

    // Default classification
    std::string err_str = error.what();
    std::transform(err_str.begin(), err_str.end(), err_str.begin(), ::tolower);

    if (err_str.find("rate limit") != std::string::npos || err_str.find("429") != std::string::npos) {
        return ErrorClass::RateLimit;
    } else if (err_str.find("timeout") != std::string::npos || err_str.find("timed out") != std::string::npos) {
        return ErrorClass::Timeout;
    } else if (err_str.find("500") != std::string::npos || err_str.find("502") != std::string::npos || err_str.find("503") != std::string::npos) {
        return ErrorClass::ServerError;
    } else if (err_str.find("400") != std::string::npos || err_str.find("401") != std::string::npos ||
               err_str.find("403") != std::string::npos || err_str.find("404") != std::string::npos) {
        return ErrorClass::ClientError;
    }

    return ErrorClass::Unknown;
}

ErrorStrategy EnhancedRetryDecorator::get_strategy(ErrorClass error_class) const {
    auto it = config_.error_strategies.find(error_class);
    if (it != config_.error_strategies.end()) {
        return it->second;
    }

    // Default strategy
    return ErrorStrategy{
        error_class,
        config_.max_attempts,
        config_.initial_backoff,
        config_.max_backoff,
        config_.backoff_multiplier,
        true
    };
}

std::chrono::milliseconds EnhancedRetryDecorator::calculate_backoff(
    std::chrono::milliseconds base_backoff,
    int attempt
) {
    double base_ms = base_backoff.count();
    std::uniform_real_distribution<> dis(0.0, 1.0);

    double jittered_ms;
    switch (config_.jitter_type) {
        case JitterType::None:
            jittered_ms = base_ms;
            break;

        case JitterType::Full:
            jittered_ms = dis(rng_) * base_ms;
            metrics_.total_jitter_added.fetch_add(
                (base_ms - jittered_ms) / 1000.0,
                std::memory_order_relaxed
            );
            break;

        case JitterType::Equal: {
            double min_backoff = base_ms * config_.jitter_min_ratio;
            jittered_ms = min_backoff + dis(rng_) * (base_ms - min_backoff);
            metrics_.total_jitter_added.fetch_add(
                (base_ms - jittered_ms) / 1000.0,
                std::memory_order_relaxed
            );
            break;
        }

        case JitterType::Decorrelated: {
            if (attempt == 1) {
                jittered_ms = base_ms;
            } else {
                auto previous = calculate_backoff(base_backoff, attempt - 1);
                double previous_ms = previous.count();
                jittered_ms = dis(rng_) * previous_ms * 3.0 + base_ms;
                if (jittered_ms > config_.max_backoff.count()) {
                    jittered_ms = config_.max_backoff.count();
                }
            }
            break;
        }

        default:
            jittered_ms = base_ms;
    }

    return std::chrono::milliseconds(static_cast<int64_t>(jittered_ms));
}

bool EnhancedRetryDecorator::check_budget(double cost) {
    if (!config_.enable_budget) {
        return true;
    }

    std::lock_guard<std::mutex> lock(budget_.mutex);

    // Reset window if hour has passed
    auto now = std::chrono::steady_clock::now();
    if (std::chrono::duration_cast<std::chrono::hours>(now - budget_.window_start).count() >= 1) {
        budget_.current_cost = 0.0;
        budget_.retry_count = 0;
        budget_.window_start = now;
    }

    // Check cost budget
    if (budget_.current_cost + cost > budget_.max_cost) {
        metrics_.budget_exceeded_count++;
        return false;
    }

    // Check retry count budget
    if (budget_.retry_count >= budget_.max_retries_per_hour) {
        metrics_.budget_exceeded_count++;
        return false;
    }

    return true;
}

bool EnhancedRetryDecorator::check_backpressure() {
    if (!config_.enable_backpressure) {
        return false;
    }

    std::lock_guard<std::mutex> lock(metrics_.mutex);

    if (metrics_.recent_results.size() < static_cast<size_t>(config_.backpressure_window)) {
        return false;
    }

    // Calculate failure rate
    int failures = 0;
    for (bool success : metrics_.recent_results) {
        if (!success) {
            failures++;
        }
    }

    double failure_rate = static_cast<double>(failures) / metrics_.recent_results.size();

    if (failure_rate > config_.backpressure_threshold) {
        metrics_.backpressure_detected++;
        return true;
    }

    return false;
}

Message EnhancedRetryDecorator::process(const Message& message) {
    std::exception_ptr last_error;
    ErrorClass error_class = ErrorClass::Unknown;
    ErrorStrategy strategy = get_strategy(error_class);

    for (int attempt = 1; attempt <= config_.max_attempts; ++attempt) {
        metrics_.total_attempts++;

        // Check budget before attempt
        if (config_.enable_budget && config_.cost_tracker) {
            double estimated_cost = config_.cost_tracker(message);
            if (!check_budget(estimated_cost)) {
                throw std::runtime_error("Retry budget exceeded");
            }
        }

        // Check backpressure
        if (check_backpressure()) {
            std::this_thread::sleep_for(std::chrono::seconds(5));
        }

        try {
            // Process message
            Message response = agent_->process(message);

            // Success
            if (attempt == 1) {
                metrics_.successful_first_attempt++;
            } else {
                metrics_.successful_on_retry++;
            }

            {
                std::lock_guard<std::mutex> lock(metrics_.mutex);
                metrics_.recent_results.push_back(true);
                if (metrics_.recent_results.size() > static_cast<size_t>(config_.backpressure_window)) {
                    metrics_.recent_results.erase(metrics_.recent_results.begin());
                }
            }

            // Track cost
            if (config_.enable_budget && config_.cost_tracker) {
                double cost = config_.cost_tracker(message);
                std::lock_guard<std::mutex> lock(budget_.mutex);
                budget_.current_cost += cost;
            }

            return response;
        } catch (const std::exception& e) {
            // Failure
            last_error = std::current_exception();

            // Track failure for backpressure
            {
                std::lock_guard<std::mutex> lock(metrics_.mutex);
                metrics_.recent_results.push_back(false);
                if (metrics_.recent_results.size() > static_cast<size_t>(config_.backpressure_window)) {
                    metrics_.recent_results.erase(metrics_.recent_results.begin());
                }
            }

            // Classify error
            error_class = classify_error(e);
            {
                std::lock_guard<std::mutex> lock(metrics_.mutex);
                metrics_.error_class_counts[error_class]++;
            }

            // Get strategy for error class
            strategy = get_strategy(error_class);

            // Check if should retry
            if (!strategy.should_retry) {
                metrics_.failed_after_retries++;
                std::rethrow_exception(last_error);
            }

            // Check if exceeded max attempts for this error class
            if (attempt >= strategy.max_attempts) {
                break;
            }

            // Track retry
            metrics_.total_retries++;
            {
                std::lock_guard<std::mutex> lock(budget_.mutex);
                budget_.retry_count++;
            }

            // Calculate backoff with jitter
            auto base_backoff_ms = std::chrono::milliseconds(
                static_cast<int64_t>(
                    strategy.initial_backoff.count() *
                    std::pow(strategy.backoff_multiplier, attempt - 1)
                )
            );
            if (base_backoff_ms > strategy.max_backoff) {
                base_backoff_ms = strategy.max_backoff;
            }
            auto backoff = calculate_backoff(base_backoff_ms, attempt);

            // Sleep with backoff
            std::this_thread::sleep_for(backoff);
        }
    }

    // All attempts failed
    metrics_.failed_after_retries++;
    if (last_error) {
        std::rethrow_exception(last_error);
    }
    throw std::runtime_error("Max retry attempts exceeded");
}

EnhancedRetryMetrics EnhancedRetryDecorator::get_metrics() const {
    return metrics_;
}

} // namespace infrastructure
} // namespace agenkit

File: agenkit-zig/src/infrastructure/retry_enhanced.zig
// Enhanced retry logic with:
// - Multiple jitter types (Full, Equal, Decorrelated)
// - Per-error-type retry strategies
// - Budget awareness (cost and count limits)
// - Backpressure detection
// - Detailed metrics

const std = @import("std");
const agksync = @import("../sync_compat.zig");
const agktime = @import("../time_compat.zig");
const Agent = @import("../agent.zig").Agent;
const Message = @import("../message.zig").Message;

/// Jitter types for retry backoff.
pub const JitterType = enum {
    none,
    full,
    equal,
    decorrelated,
};

/// Error classification for retry strategies.
pub const ErrorClass = enum {
    transient,
    rate_limit,
    timeout,
    server_error,
    client_error,
    unknown,

    pub fn asString(self: ErrorClass) []const u8 {
        return switch (self) {
            .transient => "transient",
            .rate_limit => "rate_limit",
            .timeout => "timeout",
            .server_error => "server_error",
            .client_error => "client_error",
            .unknown => "unknown",
        };
    }
};

/// Retry strategy for specific error class.
pub const ErrorStrategy = struct {
    error_class: ErrorClass,
    max_retries: usize,
    initial_delay_ms: u64,
    max_delay_ms: u64,
    multiplier: f64,
    should_retry: bool,
};

/// Retry budget to limit costs.
pub const RetryBudget = struct {
    max_cost: f64,
    current_cost: f64,
    max_retries_per_hour: u64,
    retry_count: u64,
    window_start: i64,
};

/// Enhanced retry configuration.
pub const EnhancedRetryConfig = struct {
    // Basic retry settings
    max_retries: usize,
    initial_delay_ms: u64,
    max_delay_ms: u64,
    multiplier: f64,

    // Jitter settings
    jitter_type: JitterType,
    jitter_min_ratio: f64,

    // Error-specific strategies
    error_strategies: std.AutoHashMap(ErrorClass, ErrorStrategy),

    // Budget settings
    enable_budget: bool,
    max_cost_per_hour: f64,
    max_retries_per_hour: u64,

    // Backpressure detection
    enable_backpressure: bool,
    backpressure_threshold: f64,
    backpressure_window: usize,

    pub fn default(allocator: std.mem.Allocator) !EnhancedRetryConfig {
        var error_strategies = std.AutoHashMap(ErrorClass, ErrorStrategy).init(allocator);

        try error_strategies.put(.transient, ErrorStrategy{
            .error_class = .transient,
            .max_retries = 5,
            .initial_delay_ms = 100,
            .max_delay_ms = 5000,
            .multiplier = 2.0,
            .should_retry = true,
        });

        try error_strategies.put(.rate_limit, ErrorStrategy{
            .error_class = .rate_limit,
            .max_retries = 10,
            .initial_delay_ms = 60000,
            .max_delay_ms = 300000,
            .multiplier = 1.5,
            .should_retry = true,
        });

        try error_strategies.put(.timeout, ErrorStrategy{
            .error_class = .timeout,
            .max_retries = 3,
            .initial_delay_ms = 2000,
            .max_delay_ms = 30000,
            .multiplier = 2.0,
            .should_retry = true,
        });

        try error_strategies.put(.server_error, ErrorStrategy{
            .error_class = .server_error,
            .max_retries = 3,
            .initial_delay_ms = 5000,
            .max_delay_ms = 60000,
            .multiplier = 2.0,
            .should_retry = true,
        });

        try error_strategies.put(.client_error, ErrorStrategy{
            .error_class = .client_error,
            .max_retries = 1,
            .initial_delay_ms = 0,
            .max_delay_ms = 0,
            .multiplier = 1.0,
            .should_retry = false,
        });

        return EnhancedRetryConfig{
            .max_retries = 3,
            .initial_delay_ms = 1000,
            .max_delay_ms = 30000,
            .multiplier = 2.0,
            .jitter_type = .full,
            .jitter_min_ratio = 0.5,
            .error_strategies = error_strategies,
            .enable_budget = false,
            .max_cost_per_hour = 100.0,
            .max_retries_per_hour = 1000,
            .enable_backpressure = true,
            .backpressure_threshold = 0.5,
            .backpressure_window = 100,
        };
    }

    pub fn deinit(self: *EnhancedRetryConfig) void {
        self.error_strategies.deinit();
    }
};

/// Enhanced retry metrics.
pub const EnhancedRetryMetrics = struct {
    allocator: std.mem.Allocator,
    total_attempts: u64,
    successful_first_attempt: u64,
    successful_on_retry: u64,
    failed_after_retries: u64,
    total_retries: u64,
    total_jitter_added: f64,
    budget_exceeded_count: u64,
    backpressure_detected: u64,
    error_class_counts: std.AutoHashMap(ErrorClass, u64),
    recent_results: std.ArrayList(bool),

    pub fn init(allocator: std.mem.Allocator, _: usize) EnhancedRetryMetrics {
        return .{
            .allocator = allocator,
            .total_attempts = 0,
            .successful_first_attempt = 0,
            .successful_on_retry = 0,
            .failed_after_retries = 0,
            .total_retries = 0,
            .total_jitter_added = 0.0,
            .budget_exceeded_count = 0,
            .backpressure_detected = 0,
            .error_class_counts = std.AutoHashMap(ErrorClass, u64).init(allocator),
            .recent_results = std.ArrayList(bool).empty,
        };
    }

    pub fn deinit(self: *EnhancedRetryMetrics) void {
        self.error_class_counts.deinit();
        self.recent_results.deinit();
    }
};

/// Enhanced retry decorator wraps an agent with enhanced retry logic.
pub const EnhancedRetryDecorator = struct {
    allocator: std.mem.Allocator,
    agent: Agent,
    config: EnhancedRetryConfig,
    metrics: EnhancedRetryMetrics,
    budget: RetryBudget,
    mutex: agksync.Mutex,

    pub fn init(
        allocator: std.mem.Allocator,
        agent: Agent,
        config: EnhancedRetryConfig,
    ) EnhancedRetryDecorator {
        return .{
            .allocator = allocator,
            .agent = agent,
            .config = config,
            .metrics = EnhancedRetryMetrics.init(allocator, config.backpressure_window),
            .budget = RetryBudget{
                .max_cost = config.max_cost_per_hour,
                .current_cost = 0.0,
                .max_retries_per_hour = config.max_retries_per_hour,
                .retry_count = 0,
                .window_start = agktime.milliTimestamp(),
            },
            .mutex = agksync.Mutex{},
        };
    }

    pub fn deinit(self: *EnhancedRetryDecorator) void {
        self.metrics.deinit();
    }

    pub fn name(self: *const EnhancedRetryDecorator) []const u8 {
        return self.agent.name();
    }

    pub fn capabilities(self: *const EnhancedRetryDecorator, allocator: std.mem.Allocator) ![][]const u8 {
        return self.agent.capabilities(allocator);
    }

    fn classifyError(self: *EnhancedRetryDecorator, err: anyerror) ErrorClass {
        _ = self;
        const err_name = @errorName(err);

        if (std.mem.indexOf(u8, err_name, "RateLimit") != null or
            std.mem.indexOf(u8, err_name, "429") != null)
        {
            return .rate_limit;
        } else if (std.mem.indexOf(u8, err_name, "Timeout") != null or
            std.mem.indexOf(u8, err_name, "TimedOut") != null)
        {
            return .timeout;
        } else if (std.mem.indexOf(u8, err_name, "500") != null or
            std.mem.indexOf(u8, err_name, "502") != null or
            std.mem.indexOf(u8, err_name, "503") != null)
        {
            return .server_error;
        } else if (std.mem.indexOf(u8, err_name, "400") != null or
            std.mem.indexOf(u8, err_name, "401") != null or
            std.mem.indexOf(u8, err_name, "403") != null or
            std.mem.indexOf(u8, err_name, "404") != null)
        {
            return .client_error;
        }

        return .unknown;
    }

    fn getStrategy(self: *EnhancedRetryDecorator, error_class: ErrorClass) ErrorStrategy {
        if (self.config.error_strategies.get(error_class)) |strategy| {
            return strategy;
        }

        return ErrorStrategy{
            .error_class = error_class,
            .max_retries = self.config.max_retries,
            .initial_delay_ms = self.config.initial_delay_ms,
            .max_delay_ms = self.config.max_delay_ms,
            .multiplier = self.config.multiplier,
            .should_retry = true,
        };
    }

    fn calculateBackoff(self: *EnhancedRetryDecorator, base_backoff_ms: u64, attempt: usize) u64 {
        const base_ms = @as(f64, @floatFromInt(base_backoff_ms));
        var prng = std.Random.DefaultPrng.init(@intCast(agktime.milliTimestamp()));
        const rand = prng.random();

        const jittered_ms: f64 = switch (self.config.jitter_type) {
            .none => base_ms,
            .full => rand.float(f64) * base_ms,
            .equal => blk: {
                const min_backoff = base_ms * self.config.jitter_min_ratio;
                break :blk min_backoff + rand.float(f64) * (base_ms - min_backoff);
            },
            .decorrelated => blk: {
                if (attempt == 1) {
                    break :blk base_ms;
                }
                const previous = self.calculateBackoff(base_backoff_ms, attempt - 1);
                const previous_ms = @as(f64, @floatFromInt(previous));
                var jittered = rand.float(f64) * previous_ms * 3.0 + base_ms;
                const max_ms = @as(f64, @floatFromInt(self.config.max_delay_ms));
                if (jittered > max_ms) {
                    jittered = max_ms;
                }
                break :blk jittered;
            },
        };

        return @intFromFloat(jittered_ms);
    }

    fn checkBudget(self: *EnhancedRetryDecorator, cost: f64) bool {
        if (!self.config.enable_budget) {
            return true;
        }

        self.mutex.lock();
        defer self.mutex.unlock();

        // Reset window if hour has passed
        const hour_in_ms = 3600000;
        if (agktime.milliTimestamp() - self.budget.window_start > hour_in_ms) {
            self.budget.current_cost = 0.0;
            self.budget.retry_count = 0;
            self.budget.window_start = agktime.milliTimestamp();
        }

        // Check cost budget
        if (self.budget.current_cost + cost > self.budget.max_cost) {
            self.metrics.budget_exceeded_count += 1;
            return false;
        }

        // Check retry count budget
        if (self.budget.retry_count >= self.budget.max_retries_per_hour) {
            self.metrics.budget_exceeded_count += 1;
            return false;
        }

        return true;
    }

    fn checkBackpressure(self: *EnhancedRetryDecorator) bool {
        if (!self.config.enable_backpressure) {
            return false;
        }

        self.mutex.lock();
        defer self.mutex.unlock();

        if (self.metrics.recent_results.items.len < self.config.backpressure_window) {
            return false;
        }

        // Calculate failure rate
        var failures: usize = 0;
        for (self.metrics.recent_results.items) |success| {
            if (!success) {
                failures += 1;
            }
        }

        const failure_rate = @as(f64, @floatFromInt(failures)) / @as(f64, @floatFromInt(self.metrics.recent_results.items.len));

        if (failure_rate > self.config.backpressure_threshold) {
            self.metrics.backpressure_detected += 1;
            return true;
        }

        return false;
    }

    pub fn process(self: *EnhancedRetryDecorator, message: Message) !Message {
        var last_error: ?anyerror = null;
        var error_class: ErrorClass = .unknown;
        var strategy = self.getStrategy(error_class);

        var attempt: usize = 1;
        while (attempt <= self.config.max_retries) : (attempt += 1) {
            {
                self.mutex.lock();
                self.metrics.total_attempts += 1;
                self.mutex.unlock();
            }

            // Check budget before attempt
            if (self.config.enable_budget) {
                if (!self.checkBudget(0.0)) {
                    return error.RetryBudgetExceeded;
                }
            }

            // Check backpressure
            if (self.checkBackpressure()) {
                agktime.sleep(5000 * std.time.ns_per_ms);
            }

            // Process message
            const result = self.agent.process(message);

            if (result) |response| {
                // Success
                self.mutex.lock();
                if (attempt == 1) {
                    self.metrics.successful_first_attempt += 1;
                } else {
                    self.metrics.successful_on_retry += 1;
                }
                try self.metrics.recent_results.append(true);
                if (self.metrics.recent_results.items.len > self.config.backpressure_window) {
                    _ = self.metrics.recent_results.orderedRemove(0);
                }
                self.mutex.unlock();

                return response;
            } else |err| {
                // Failure
                last_error = err;

                // Track failure for backpressure
                {
                    self.mutex.lock();
                    try self.metrics.recent_results.append(false);
                    if (self.metrics.recent_results.items.len > self.config.backpressure_window) {
                        _ = self.metrics.recent_results.orderedRemove(0);
                    }
                    self.mutex.unlock();
                }

                // Classify error
                error_class = self.classifyError(err);
                {
                    self.mutex.lock();
                    const entry = try self.metrics.error_class_counts.getOrPut(error_class);
                    if (entry.found_existing) {
                        entry.value_ptr.* += 1;
                    } else {
                        entry.value_ptr.* = 1;
                    }
                    self.mutex.unlock();
                }

                // Get strategy for error class
                strategy = self.getStrategy(error_class);

                // Check if should retry
                if (!strategy.should_retry) {
                    self.mutex.lock();
                    self.metrics.failed_after_retries += 1;
                    self.mutex.unlock();
                    return error.NonRetryableError;
                }

                // Check if exceeded max attempts for this error class
                if (attempt >= strategy.max_retries) {
                    break;
                }

                // Track retry
                {
                    self.mutex.lock();
                    self.metrics.total_retries += 1;
                    self.budget.retry_count += 1;
                    self.mutex.unlock();
                }

                // Calculate backoff with jitter
                const exp = @as(f64, @floatFromInt(attempt - 1));
                const base_backoff_ms_float = @as(f64, @floatFromInt(strategy.initial_delay_ms)) * std.math.pow(f64, strategy.multiplier, exp);
                var base_backoff_ms = @as(u64, @intFromFloat(base_backoff_ms_float));
                if (base_backoff_ms > strategy.max_delay_ms) {
                    base_backoff_ms = strategy.max_delay_ms;
                }
                const backoff_ms = self.calculateBackoff(base_backoff_ms, attempt);

                // Sleep with backoff
                agktime.sleep(backoff_ms * std.time.ns_per_ms);
            }
        }

        // All attempts failed
        self.mutex.lock();
        self.metrics.failed_after_retries += 1;
        self.mutex.unlock();

        return last_error orelse error.MaxRetryAttemptsExceeded;
    }

    pub fn getMetrics(self: *EnhancedRetryDecorator) EnhancedRetryMetrics {
        self.mutex.lock();
        defer self.mutex.unlock();
        return self.metrics;
    }
};
